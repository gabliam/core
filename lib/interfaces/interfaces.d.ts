/// <reference types="express" />
import * as express from 'express';
import { Gabliam } from '../gabliam';
import { interfaces } from "inversify";
export interface Controller {
}
export interface HandlerDecorator {
    (target: any, key: string, value: any): void;
}
export interface ConfigFunction {
    (app: express.Application): void;
}
export interface RoutingConfig {
    rootPath: string;
}
export declare type ModuleFunction = (framework: Gabliam) => Promise<void>;
export interface Config {
}
export interface ConfigRegistry {
    order: number;
    id: interfaces.ServiceIdentifier<any>;
}
