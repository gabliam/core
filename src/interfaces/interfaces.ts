import { interfaces } from 'inversify';
import { Registry } from '../registry';
import { ValueValidator, PluginMetadata } from './metadata';

/**
 * Config for gabliam
 */
export interface GabliamConfig {
  /**
   * Path to scan
   */
  scanPath: string;

  /**
   * Path of config
   */
  configPath: string;
}

/**
 * Interface of plugin constructor
 */
export interface GabliamPluginConstructor {
  new (): GabliamPlugin;
}

export interface GabliamPluginDefinition extends PluginMetadata {
  plugin: GabliamPlugin;
}

/**
 * Interface for a plugin
 */
export interface GabliamPlugin {
  build?(
    container: interfaces.Container,
    registry: Registry
  ): void | Promise<void>;

  bind?(
    container: interfaces.Container,
    registry: Registry
  ): void | Promise<void>;

  config?(
    container: interfaces.Container,
    registry: Registry,
    confInstance: any
  ): void | Promise<void>;

  start?(container: interfaces.Container, registry: Registry): Promise<void>;

  stop?(container: interfaces.Container, registry: Registry): Promise<void>;

  destroy?(container: interfaces.Container, registry: Registry): Promise<void>;
}

export type ValueExtractor = (
  path: string,
  defaultValue: any,
  validator?: ValueValidator | null
) => any;
