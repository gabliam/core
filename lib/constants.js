"use strict";
const TYPE = {
    Controller: Symbol("Controller"),
    Service: Symbol("Service"),
    Config: Symbol("Config")
};
exports.TYPE = TYPE;
const METADATA_KEY = {
    controller: "_controller",
    controllerMethod: "_controller-method",
    Bean: "_bean",
};
exports.METADATA_KEY = METADATA_KEY;
const DEFAULT_ROUTING_ROOT_PATH = "/";
exports.DEFAULT_ROUTING_ROOT_PATH = DEFAULT_ROUTING_ROOT_PATH;
