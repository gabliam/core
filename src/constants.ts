/** Type in registry */
export const TYPE = {
    Service: 'ServiceType',
    Config: 'ConfigType',
};

export const METADATA_KEY = {
    bean: '_bean',
    value: '_value',
    register: '_register',
    scan: '_scan',
};

export const ORDER_CONFIG = {
    Core: 50,
    Plugin: 100,
    Config: 150,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');
