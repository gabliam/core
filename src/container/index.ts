import { Container } from 'inversify';
import { makeLoggerMiddleware } from 'inversify-logger-middleware';
import { makeValueMiddleware } from './valueMiddleware';

/**
 * Create the inversify container
 */
export function createContainer(): Container {
    let container = new Container();

    let middlewares = [];
    if (process.env.NODE_ENV === 'development') {
        let logger = makeLoggerMiddleware();
        middlewares.push(logger);
    }

    middlewares.push(makeValueMiddleware(container));
    container.applyMiddleware(...middlewares);
    return container;
}

export { Container };
