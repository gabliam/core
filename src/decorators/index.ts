export * from './service';
export * from './bean';
export { Config, PluginConfig } from './config';
export * from './value';
export * from './scan';
export * from './register';
