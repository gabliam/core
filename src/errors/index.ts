export * from './value-validation-error';
