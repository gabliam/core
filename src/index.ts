import * as interfaces from './interfaces';
export * from './decorators';
export * from './gabliam';
export { APP_CONFIG, CORE_CONFIG } from './constants';

export * from './registry';
export * from './errors';

export {
  injectable, tagged, named, inject, optional, unmanaged, multiInject, targetName, decorate, interfaces as inversifyInterfaces
} from 'inversify';

export { interfaces };
