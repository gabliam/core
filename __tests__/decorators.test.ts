// tslint:disable:one-line
// tslint:disable:no-unused-expression
import {
  Bean,
  Config,
  register,
  Scan,
  Service,
  Value
} from '../src/decorators';
import { METADATA_KEY, TYPE } from '../src/constants';
import { BeanMetadata, RegistryMetada, ValueMetadata } from '../src/interfaces';
import * as Joi from 'joi';

describe('Unit Test: Decorators', () => {
  describe('@Bean', () => {
    it('should add Bean metadata to a class when decorated with @Bean', () => {
      class TestBean {
        @Bean('test')
        testMethod() {}

        @Bean('test2')
        test2Method() {}
      }

      const beanMetadata: BeanMetadata[] = Reflect.getMetadata(
        METADATA_KEY.bean,
        TestBean
      );

      expect(beanMetadata).toMatchSnapshot();
    });

    it('should add Bean metadata to a class when decorated multiple times with @Bean', () => {
      class TestBean {
        @Bean('test')
        testMethod() {}

        @Bean('test2')
        @Bean('test3')
        test2Method() {}
      }

      const beanMetadata: BeanMetadata[] = Reflect.getMetadata(
        METADATA_KEY.bean,
        TestBean
      );
      expect(beanMetadata).toMatchSnapshot();
    });
  }); // end describe @Bean

  describe('@Config', () => {
    it('should add Registry and config metadata to a class when decorated with @Config', () => {
      @Config()
      class TestBean {}

      const registryMetadata: RegistryMetada = Reflect.getMetadata(
        METADATA_KEY.register,
        TestBean
      );

      const configMetadata: boolean | undefined = Reflect.getMetadata(
        METADATA_KEY.config,
        TestBean
      );

      expect(configMetadata).toMatchSnapshot();
      expect(registryMetadata).toMatchSnapshot();
    });

    it('should fail when decorated multiple times with @Config', () => {
      expect(function() {
        @Config()
        @Config()
        class TestBean {}

        new TestBean();
      }).toThrowError();
    });
  }); // end describe @Config

  describe('@Register', () => {
    it('should add Registry metadata to a class when decorated with @Register', () => {
      @register(TYPE.Config, {})
      class TestBean {}

      const registryMetadata: RegistryMetada = Reflect.getMetadata(
        METADATA_KEY.register,
        TestBean
      );

      expect(registryMetadata).toMatchSnapshot();
    });

    it('should fail when decorated multiple times with @Register', () => {
      expect(function() {
        @register(TYPE.Config, {})
        @register(TYPE.Service, {})
        class TestBean {}

        new TestBean();
      }).toThrowError();
    });
  }); // end describe @Register

  describe('@Scan', () => {
    it('should add Scan metadata to a class when decorated with @Scan', () => {
      @Scan('__tests/')
      class TestBean {}

      const scanMetadata: RegistryMetada = Reflect.getMetadata(
        METADATA_KEY.scan,
        TestBean
      );

      expect(scanMetadata).toMatchSnapshot();
    });

    it('should add Scan metadata to a class when decorated multiple times with @Scan', () => {
      @Scan('__tests/')
      @Scan(`__tests/otherFolder`)
      class TestBean {}

      const scanMetadata: string[] = Reflect.getMetadata(
        METADATA_KEY.scan,
        TestBean
      );
      expect(scanMetadata).toMatchSnapshot();
    });
  }); // end describe @Scan

  describe('@Service', () => {
    it('should add Service metadata to a class when decorated with @Service', () => {
      @Service()
      class TestBean {}

      const registryMetadata: RegistryMetada = Reflect.getMetadata(
        METADATA_KEY.register,
        TestBean
      );

      const serviceMetadata: boolean | undefined = Reflect.getMetadata(
        METADATA_KEY.service,
        TestBean
      );
      expect(serviceMetadata).toMatchSnapshot();
      expect(registryMetadata).toMatchSnapshot();
    });

    it(`should add Service metadata to a class when decorated with @Service('Test)`, () => {
      @Service('Test')
      class TestBean {}

      const registryMetadata: RegistryMetada = Reflect.getMetadata(
        METADATA_KEY.register,
        TestBean
      );

      const serviceMetadata: boolean = Reflect.getMetadata(
        METADATA_KEY.service,
        TestBean
      );

      expect(serviceMetadata).toMatchSnapshot();
      expect(registryMetadata).toMatchSnapshot();
    });

    it('should fail when decorated multiple times with @Service', () => {
      expect(function() {
        @Service()
        @Service()
        class TestBean {}

        new TestBean();
      }).toThrowError();
    });
  }); // end describe @Service

  describe('@Value', () => {
    describe('@Value(options: ValueOptions)', () => {
      it('should add Value metadata to a class when decorated with @Value(options: ValueOptions)', () => {
        class TestBean {
          @Value({ path: 'application.name' })
          name: string;
          @Value({ path: 'application.surname', validator: Joi.string() })
          surname: string;
          @Value({
            path: 'application.surname',
            validator: { schema: Joi.string() }
          })
          firstname: string;
          @Value({
            path: 'application.postalcode',
            validator: {
              schema: Joi.string().required(),
              customErrorMsg: 'Error'
            }
          })
          postalcode: string;
          @Value({
            path: 'application.address',
            validator: Joi.string().required()
          })
          address: string;
        }

        const valueMetadata: ValueMetadata[] = Reflect.getMetadata(
          METADATA_KEY.value,
          TestBean
        );
        expect(valueMetadata).toMatchSnapshot();
      });
    }); // end @Value(options: ValueOptions)
  }); // end describe @Value
});
