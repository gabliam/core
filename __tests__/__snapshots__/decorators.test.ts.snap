// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Bean should add Bean metadata to a class when decorated multiple times with @Bean 1`] = `
Array [
  Object {
    "id": "test",
    "key": "testMethod",
  },
  Object {
    "id": "test3",
    "key": "test2Method",
  },
  Object {
    "id": "test2",
    "key": "test2Method",
  },
]
`;

exports[`@Bean should add Bean metadata to a class when decorated with @Bean 1`] = `
Array [
  Object {
    "id": "test",
    "key": "testMethod",
  },
  Object {
    "id": "test2",
    "key": "test2Method",
  },
]
`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config 1`] = `true`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 150,
    },
    "target": [Function],
  },
}
`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config(100) 1`] = `true`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config(100) 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 100,
    },
    "target": [Function],
  },
}
`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig 1`] = `true`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 50,
    },
    "target": [Function],
  },
}
`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig(100) 1`] = `true`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig(100) 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 100,
    },
    "target": [Function],
  },
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin('TestPlugin') 1`] = `
Object {
  "dependencies": Array [],
  "name": "TestPlugin",
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin() 1`] = `
Object {
  "dependencies": Array [],
  "name": "TestBean",
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ dependencies: ['TestPlugin'] }) 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "TestPlugin",
      "order": "after",
    },
  ],
  "name": "TestBean",
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin' }) 1`] = `
Object {
  "dependencies": Array [],
  "name": "TestPlugin",
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin', dependencies: ['TestPlugin2'] }) 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "TestPlugin2",
      "order": "after",
    },
  ],
  "name": "TestPlugin",
}
`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig 1`] = `true`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 100,
    },
    "target": [Function],
  },
}
`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig(100) 1`] = `true`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig(100) 2`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "options": Object {
      "order": 100,
    },
    "target": [Function],
  },
}
`;

exports[`@Register should add Registry metadata to a class when decorated with @Register 1`] = `
Object {
  "type": "ConfigType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@Scan should add Scan metadata to a class when decorated multiple times with @Scan 1`] = `
Array [
  "<PROJECT_ROOT>/__tests__/otherFolder",
  "<PROJECT_ROOT>/__tests__",
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan() 1`] = `
Array [
  "<PROJECT_ROOT>/__tests__",
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(__dirname) 1`] = `
Array [
  "<PROJECT_ROOT>/__tests__",
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(relativePath) 1`] = `
Array [
  "<PROJECT_ROOT>/__tests__/fixtures/loader",
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(relativePath) 2 1`] = `
Array [
  "<PROJECT_ROOT>/src",
]
`;

exports[`@Service should add Service metadata to a class when decorated with @Service 1`] = `true`;

exports[`@Service should add Service metadata to a class when decorated with @Service 2`] = `
Object {
  "type": "ServiceType",
  "value": Object {
    "id": [Function],
    "target": [Function],
  },
}
`;

exports[`@Service should add Service metadata to a class when decorated with @Service('Test) 1`] = `true`;

exports[`@Service should add Service metadata to a class when decorated with @Service('Test) 2`] = `
Object {
  "type": "ServiceType",
  "value": Object {
    "id": "Test",
    "target": [Function],
  },
}
`;

exports[`@Value @Value(options: ValueOptions) should add Value metadata to a class when decorated with @Value(options: ValueOptions) 1`] = `
Array [
  Object {
    "key": "name",
    "path": "application.name",
    "target": TestBean {},
    "validator": null,
  },
  Object {
    "key": "surname",
    "path": "application.surname",
    "target": TestBean {},
    "validator": Object {
      "schema":  {
        "_description": null,
        "_examples": Array [],
        "_flags": Object {},
        "_inner": Object {},
        "_invalids": Set {
          "_set": Array [
            "",
          ],
        },
        "_meta": Array [],
        "_notes": Array [],
        "_refs": Array [],
        "_settings": null,
        "_tags": Array [],
        "_tests": Array [],
        "_type": "string",
        "_unit": null,
        "_valids": Set {
          "_set": Array [],
        },
        "isJoi": true,
      },
      "throwError": true,
    },
  },
  Object {
    "key": "firstname",
    "path": "application.surname",
    "target": TestBean {},
    "validator": Object {
      "schema":  {
        "_description": null,
        "_examples": Array [],
        "_flags": Object {},
        "_inner": Object {},
        "_invalids": Set {
          "_set": Array [
            "",
          ],
        },
        "_meta": Array [],
        "_notes": Array [],
        "_refs": Array [],
        "_settings": null,
        "_tags": Array [],
        "_tests": Array [],
        "_type": "string",
        "_unit": null,
        "_valids": Set {
          "_set": Array [],
        },
        "isJoi": true,
      },
      "throwError": true,
    },
  },
  Object {
    "key": "postalcode",
    "path": "application.postalcode",
    "target": TestBean {},
    "validator": Object {
      "customErrorMsg": "Error",
      "schema":  {
        "_baseType": undefined,
        "_description": null,
        "_examples": Array [],
        "_flags": Object {
          "presence": "required",
        },
        "_inner": Object {},
        "_invalids": Set {
          "_set": Array [
            "",
          ],
        },
        "_meta": Array [],
        "_notes": Array [],
        "_refs": Array [],
        "_settings": null,
        "_tags": Array [],
        "_tests": Array [],
        "_type": "string",
        "_unit": null,
        "_valids": Set {
          "_set": Array [],
        },
        "isJoi": true,
      },
      "throwError": true,
    },
  },
  Object {
    "key": "address",
    "path": "application.address",
    "target": TestBean {},
    "validator": Object {
      "schema":  {
        "_baseType": undefined,
        "_description": null,
        "_examples": Array [],
        "_flags": Object {
          "presence": "required",
        },
        "_inner": Object {},
        "_invalids": Set {
          "_set": Array [
            "",
          ],
        },
        "_meta": Array [],
        "_notes": Array [],
        "_refs": Array [],
        "_settings": null,
        "_tags": Array [],
        "_tests": Array [],
        "_type": "string",
        "_unit": null,
        "_valids": Set {
          "_set": Array [],
        },
        "isJoi": true,
      },
      "throwError": true,
    },
  },
]
`;
